### ISSUE TITLE
recognize fails with URL-encoded params

### ISSUE THREAD

**Issue by jamesarosen** [tildeio/route-recognizer#71]
Failing test:
```js
test("A dynamic route with encoded params", function() {
  var handler = {};
  var router = new RouteRecognizer();
  router.add([{ path: "/foo/:bar", handler: handler }]);

  // Passes:
  resultsMatch(router.recognize("/foo/bar%20baz"), [{ handler: handler, params: { bar: "bar baz" }, isDynamic: true }]);

  // Fails:
  resultsMatch(router.recognize("/foo/bar%3fbaz"), [{ handler: handler, params: { bar: "bar?baz" }, isDynamic: true }]);

  // Fails:
  resultsMatch(router.recognize("/foo/bar%23baz"), [{ handler: handler, params: { bar: "bar#baz" }, isDynamic: true }]);
});
```

**Comment by jamesarosen** [comment]
Hmm. The following test passes:
```js
test("A dynamic route with encoded params", function() {
  var handler = {};
  var router = new RouteRecognizer();
  router.add([{ path: "/foo/:bar", handler: handler }]);

  equal(router.recognize("/foo/bar%20baz")[0].params.bar, "bar baz");
  equal(router.recognize("/foo/bar%23baz")[0].params.bar, "bar%23baz");
  equal(router.recognize("/foo/bar%3fbaz")[0].params.bar, "bar%3fbaz");
});
```
Perhaps this isn't the source of my problem (which is that `PUT /foo/bar %3f baz` shows up as `PUT /foo/bar ` in Mirage).

**Comment by jamesarosen** [comment]
OK, I don't have a failing test for this, but I do have a [failing test for Pretender](https://github.com/pretenderjs/pretender/issues/124).

Changing
```js
path = decodeURI(path);
```
to
```js
path = decodeURIComponent(path);
```
in `recognize` fixes that failure on Chrome, but not PhantomJS.

**Comment by seanjohnson08** [comment]
I have a related issue:

In the app that I am currently maintaining, we have some external URLs that are coming in that are improperly escaped. These URLs have been out in the wild now for over 5 years, and there's not a whole lot we can do to change them.

Here's an example of a poorly encoded URL: /compose-message/Parab%E9ns%20pelo%20seu%20novo%20cargol

The key takeaway is that "%E9" - that is the result of calling escape('é'). I know that the router is doing the right thing by using the URI family of functions (they are UTF compliant) but would it be at all possible to either handle these poorly escaped URLs or give me some kind of hook into which I can handle them myself?

**Comment by jmeas** [comment]
@seanjohnson08 I think providing a hook for that makes a lot of sense.

@jamesarosen any updates? Is this still an issue?

**Comment by jamesarosen** [comment]
No updates. Still very much an issue (on 0.1.9). I'd be happy to work on this if there's consensus on the correct solution. I'm wary of doing a bunch of work that gets thrown away.

**Comment by seanjohnson08** [comment]
in my case, the solution would be as simple as

```try {
  path = decodeURIComponent(path);
} catch (e) {
  path = unescape(path);
}```

But the right solution may look a bit different. I at least believe that the fact that `decodeURI` can throw should be dealt with in a more user-friendly way.

**Comment by rwjblue** [comment]
I believe that this was resolved by https://github.com/tildeio/route-recognizer/pull/91 (and included in 0.2.0).  That change will be pulled into Ember in 2.7.
